### PR TITLE
Various cleanups

### PR DESCRIPTION
**Pull request by jonsneyers**
Various cleanups all aimed at reducing the amount of unnecessary strings and debug code in libjxl.so:

1. JXL_ABORT() is often used in unreachable code paths that could only become reachable if the code would change (e.g. if the encoder would use a new blendmode for patches, or some programming error happened that caused a presumably unreachable path to become reachable). However, the compiler cannot know these are unreachable so even release builds still contain the code to print the abort message and to do the actual abort. This replaces these aborts with JXL_NOT_REACHABLE() which is a new macro that keeps the old behavior for debug builds (print message and abort), but in release builds it simply signals JXL_UNREACHABLE to inform the compiler that this is in fact an unreachable path.

2. Butteraugli had some code for generic convolutions that aren't actually used (and there was a printf warning). Replaced it with JXL_NOT_REACHABLE().

3. Removed deprecated dead code from `butteraugli.*`

4. Adaptive quantization debug stuff (dumping debug images and printf'ing stuff) ended up getting compiled in even in release builds. Putting it behind a compile flag `JXL_DEBUG_ADAPTIVE_QUANTIZATION`.

5. There are a bunch of printf strings in enc_aux_out that also end up in libjxl.so. Putting these behind `#if JXL_DEBUG_V_LEVEL > 0`.

### CODE REVIEW

**Comment by veluca93** [comment]
Why do we need a new macro JXL_NOT_REACHABLE? Couldn't we just use
JXL_UNREACHABLE?

**Comment by jonsneyers** [comment]
You mean change JXL_UNREACHABLE to take a message argument and also abort in debug builds?
Yes, that could also be done. I kept the existing macro like it was because perhaps there are cases where you just want to signal to the compiler that a path is unreachable, without writing a specific debug message and aborting.

**Comment by veluca93** [comment]
> You mean change JXL_UNREACHABLE to take a message argument and also abort in debug builds? Yes, that could also be done. I kept the existing macro like it was because perhaps there are cases where you just want to signal to the compiler that a path is unreachable, without writing a specific debug message and aborting.

Considering actually reaching a `__builtin_unreachable()` statement is undefined behaviour, I cannot imagine a single valid reason not to want to debug-message+abort just before that.

**Comment by veluca93** [comment]
I think it would cause less future confusion if the macros were called JXL_UNREACHABLE and JXL_UNREACHABLE_BUILTIN instead.

**Comment by jonsneyers** [comment]
That makes sense. Renamed the macros accordingly.

Also made it so the debug messages of JXL_CHECK (which can be quite long in some cases) do not end up in release builds.

File: lib/jxl/butteraugli/butteraugli.h [hunk 1]
```diff
@@ -43,49 +43,6 @@ struct ButteraugliParams {
...
-//
-// A diffvalue smaller than kButteraugliGood indicates that images can be
-// observed as the same image.
-// diffvalue larger than kButteraugliBad indicates that a difference between
-// the images can be observed.
-// A diffvalue between kButteraugliGood and kButteraugliBad indicates that
-// a subtle difference can be observed between the images.
-//
-// Returns true on success.
-bool ButteraugliInterface(const Image3F &rgb0, const Image3F &rgb1,
```
Comment (veluca93) [inline lib/jxl/butteraugli/butteraugli.h:80]: We have some internal users for this API, please leave it around.

File: lib/jxl/enc_adaptive_quantization.cc [hunk 2]
```diff
@@ -877,15 +883,15 @@ void FindBestQuantization(const ImageBundle& linear, const Image3F& opsin,
...
   }
   for (int i = 0; i < iters + 1; ++i) {
-    if (FLAGS_dump_quant_state) {
-      printf("\nQuantization field:\n");
-      for (size_t y = 0; y < quant_field.ysize(); ++y) {
-        for (size_t x = 0; x < quant_field.xsize(); ++x) {
-          printf(" %.5f", quant_field.Row(y)[x]);
-        }
-        printf("\n");
+#if JXL_DEBUG_ADAPTIVE_QUANTIZATION
```
Comment (veluca93) [inline lib/jxl/enc_adaptive_quantization.cc:886]: Let's avoid `#if` to skip compiling code and use `if (some_constexpr_boolean)` instead.

File: lib/jxl/enc_aux_out.cc [hunk 3]
```diff
@@ -100,6 +102,7 @@ void AuxOut::Assimilate(const AuxOut& victim) {
 }
 
 void AuxOut::Print(size_t num_inputs) const {
+#if JXL_DEBUG_V_LEVEL > 0
```
Comment (veluca93) [inline lib/jxl/enc_aux_out.cc:105]: Ideally, let's do the same thing here (`if (JXL_DEBUG_V_LEVEL > 0)`)

File: lib/jxl/butteraugli/butteraugli.h (hunk 1 above)
Comment (jonsneyers) [reply lib/jxl/butteraugli/butteraugli.h:80]: OK, but probably it shouldn't be in `lib/jxl/` if it isn't used by libjxl (or anything in this repo, even).
Does the deprecated function below also need to be kept or can that one be removed?

Comment (veluca93) [reply lib/jxl/butteraugli/butteraugli.h:80]: Let's leave it around for now, and discuss further with Jyrki.

**Comment by jonsneyers** [comment]
Also reverted the manual unrolling in jpeg reconstruction from https://github.com/libjxl/libjxl/pull/2534 (see https://github.com/libjxl/libjxl/pull/2534#discussion_r1221643355), since it was adding ~100kb of binary size for a relatively small speedup in a relatively niche code path (reconstructing non-progressive jpegs).

File: lib/jxl/enc_aux_out.cc (hunk 3 above)
Comment (jonsneyers) [reply lib/jxl/enc_aux_out.cc]: done

File: lib/jxl/enc_adaptive_quantization.cc (hunk 2 above)
Comment (jonsneyers) [reply lib/jxl/enc_adaptive_quantization.cc]: ok — there are probably quite a few other spots where we were already doing that, feel free to update those too :)

File: lib/jxl/butteraugli/butteraugli.h (hunk 1 above)
Comment (jonsneyers) [reply lib/jxl/butteraugli/butteraugli.h]: ok, kept butteraugli as it was except for the not-implemented-kernel-size thing.

**Comment by veluca93** [comment]
> Also reverted the manual unrolling in jpeg reconstruction from #2534 (see [#2534 (comment)](https://github.com/libjxl/libjxl/pull/2534#discussion_r1221643355)), since it was adding ~100kb of binary size for a relatively small speedup in a relatively niche code path (reconstructing non-progressive jpegs).

Can you split that in another pull request, and sent that to @szabadka ?

**Comment by jonsneyers** [comment]
> > Also reverted the manual unrolling in jpeg reconstruction from #2534 (see [#2534 (comment)](https://github.com/libjxl/libjxl/pull/2534#discussion_r1221643355)), since it was adding ~100kb of binary size for a relatively small speedup in a relatively niche code path (reconstructing non-progressive jpegs).
> 
> Can you split that in another pull request, and sent that to @szabadka ?

sure

**Review by veluca93 — APPROVED** [review]

**Pull request merged**